### PR TITLE
105 add timetime service

### LINKED ISSUE #105: Add TimeTime Service
**Short description:**

Add a new Service named TimeTime :
- This service will handle all Time relative operation

Action :
- Do Something at a date
- Do Something every X time 

Reaction : 
- No Idea for the moment

### PR DESCRIPTION
**Pull request by LucasTesnier**
# Purpose

The purpose of this pull request is to add the service Time Time.

Fixes #105 

## Added features

* The Time Time service is now available !
* Action at a date is available
* Action every X time is available

* A trigger creation / destruction is now implemented

image

**Reminded : Some major bug with Flutter app have been identified and need to be fix before the end of the Week**

### CODE REVIEW

File: server/server_app.js [hunk 1]
```diff
@@ -975,8 +1007,21 @@ app.post(
...
+            Action: true
           }
         })
+      if (TriggerInitMap[areaCreation.Action.code])
+        if (!TriggerInitMap[areaCreation.Action.code](areaCreation)) {
+          await database.prisma.UsersHasActionsReactions.delete({
+            where: { id: areaCreation.id }
+          })
+          return res.status(400).send('Please pass a valid parameter list !')
+        }
```
Comment (ZQUEMA) [inline server/server_app.js:1041]: Why do you delete smth in a creation request ?

File: server/services/timetime/actions/at_a_date.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+ * @returns the date object if paramsList is valid, false otherwise
+ */
+function parseAreaParameterToGetDate (paramsList) {
+  try {
+    const year = paramsList[0].value.split('/')[0]
+    const month = paramsList[0].value.split('/')[1]
+    const day = paramsList[0].value.split('/')[2]
+    const hour = paramsList[1].value != '' ? paramsList[1].value : '0'
+    const min = paramsList[2].value != '' ? paramsList[2].value : '0'
+    const sec = paramsList[3].value != '' ? paramsList[3].value : '0'
```
Comment (ZQUEMA) [inline server/services/timetime/actions/at_a_date.js:32]: Why does it use hard array indexes ? Can't it be like the other actions ?

File: server/services/timetime/actions/at_a_date.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    rule,
+    function (area) {
+      var currentJob = getATimeTimeJobById(area.id)
+      if (currentJob != null) currentJob.occurence -= 1
+      if (currentJob != null && currentJob.occurence <= 0)
+        destroyATimeTimeAtADate(area)
+      const parametersList = [
+        {
+          name: 'date',
+          value: area.ActionParameters[0].value,
```
Comment (ZQUEMA) [inline server/services/timetime/actions/at_a_date.js]: Why does it use hard array indexes ? Can't be like the other actions ?

File: server/services/timetime/actions/at_a_date.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+          valid: false
+        },
+        {
+          name: 'hour',
+          value: area.ActionParameters[1].value,
```
Comment (ZQUEMA) [inline server/services/timetime/actions/at_a_date.js]: Same as above

File: server/services/timetime/actions/at_a_date.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+          valid: false
+        },
+        {
+          name: 'minute',
+          value: area.ActionParameters[2].value,
```
Comment (ZQUEMA) [inline server/services/timetime/actions/at_a_date.js]: Same as above

File: server/services/timetime/actions/at_a_date.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+          valid: false
+        },
+        {
+          name: 'second',
+          value: area.ActionParameters[3].value,
```
Comment (ZQUEMA) [inline server/services/timetime/actions/at_a_date.js]: Same as above

File: server/services/timetime/actions/every_x_time.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Parse all parameters and return the date rule
+ * @param {*} parameters the parameters list
+ * @returns the date rule object if paramsList is valid, false otherwise
+ */
+function createFrequencyRule (parameters) {
+  if (parseInt(parameters[0].value) < 1) return null
+  if (parseInt(parameters[1].value) < 1) return null
+  return '*/' + parameters[0].value + ' * * * * *'
```
Comment (ZQUEMA) [inline server/services/timetime/actions/every_x_time.js:18]: Why does it use hard array indexes ? Can't be like the other actions ?

File: server/services/timetime/actions/every_x_time.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    rule,
+    function (area) {
+      var currentJob = getATimeTimeJobById(area.id)
+      if (currentJob != null) currentJob.occurence -= 1
+      if (currentJob != null && currentJob.occurence <= 0)
+        destroyATimeTimeAtX(area)
+      const parametersList = [
+        {
+          name: 'every',
+          value: area.ActionParameters[0].value,
```
Comment (ZQUEMA) [inline server/services/timetime/actions/every_x_time.js]: Why does it use hard array indexes ? Can't be like the other actions ?

File: server/services/timetime/actions/every_x_time.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+          valid: false
+        },
+        {
+          name: 'occurence',
+          value: area.ActionParameters[1].value,
```
Comment (ZQUEMA) [inline server/services/timetime/actions/every_x_time.js]: Same as above

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
Answer the questions

File: server/server_app.js (hunk 1 above)
Comment (LucasTesnier) [reply server/server_app.js:1041]: If the Trigger init failed (For some bad parameters meaning or other things) we need to delete the newly created area.

Comment (ZQUEMA) [reply server/server_app.js:1041]: If smth failled, the prisma query will send a error and won't create the service

**Review by tbellicha — DISMISSED** [review]
NTR

**Review by TomDUVAL-MAHE — APPROVED** [review]
NTR

**Review by ZQUEMA — APPROVED** [review]

**Pull request merged**